akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,246 @@
+      var buffer: PartitionedBuffer[Partition, Holder[Partition, In, Out]] = _
+
+      private val futureCB = getAsyncCallback[Holder[Partition, In, Out]] { holder =>
+        holder.outgoing match {
+          case Failure(ex) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop => failStage(ex)
+          case _ =>
+            dropCompletedThenPushIfPossible(holder.partition)
+        }
+      }
+
+      override def preStart(): Unit = {
+        buffer = new PartitionedBuffer[Partition, Holder[Partition, In, Out]](
+          linearBuffer = BufferImpl(parallelism, inheritedAttributes),
+          partitioner = _.partition,
+          makePartitionBuffer = { (p, overflowCapacity) =>
+            import akka.stream.impl.{ BoundedBuffer, ChainedBuffer }
+            val tailCapacity = (overflowCapacity - perPartition).max(1)
+
+            new ChainedBuffer[Holder[Partition, In, Out]](
+              BufferImpl(perPartition, inheritedAttributes),
+              new BoundedBuffer.DynamicQueue(tailCapacity), {
+                holder =>
+                  // this will execute when the holder moves from the tail buffer to the head buffer (viz. ready to be
+                  //  scheduled).  This might be on enqueueing into the partitioned buffer, or it might be when
+                  //  the head of this partition is dropped after being completed.
+                  try {
+                    val future = f(holder.incoming, p)
+                    holder.clearIncoming()
+
+                    future.value match {
+                      case None    => future.onComplete(holder)(akka.dispatch.ExecutionContexts.parasitic)
+                      case Some(v) =>
+                        // future already completed, so don't schedule on dispatcher
+                        holder.setOutgoing(v)
+                        v match {
+                          case Failure(ex) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop =>
+                            failStage(ex)
+
+                          case _ =>
+                            dropCompletedThenPushIfPossible(p)
+                        }
+                    }
+                  } catch {
+                    // executes if f throws, not a failed future
+                    case NonFatal(ex) =>
+                      val directive = holder.supervisionDirectiveFor(decider, ex)
+                      if (directive == Supervision.Stop) failStage(ex)
+                      else {
+                        holder.clearIncoming()
+                        if (holder.outgoing == NotYetThere) {
+                          holder.setOutgoing(Failure(ex))
+                          dropCompletedThenPushIfPossible(p)
+                        } else {
+                          failStage(
+                            new AssertionError("Should only get here if f throws, not if future was created", ex))
```

Comment:
Let's use `IllegalStateException` instead, to align with other Akka code.

Reply:
OK.  I guess `IllegalStateException` works as a signal that an assumption we're making about the state of the world doesn't hold.